### PR TITLE
VLAN overhaul

### PR DESCRIPTION
**Pull request by jukkar**
This one overhauls VLAN support to use virtual interfaces. When the VLAN support was originally created, there was no virtual interfaces so the VLAN support was implemented a bit hackish way.
This does not change user APIs related to VLANs, but the Ethernet drivers need to be changed meaning that most of the VLAN code can be removed from the driver.

Tested Ethernet drivers:
- [x] stm32, used `nucleo_f767zi`
- [x] native_sim
- [x] sam_gmac, used `sam_e70_xplained`
- [ ] enc28j60
- [x] mcux, used `frdm-k64f` and `mimxrt1050_evk`
- [x] nxp_s32_gmac
- [x] nxp_enet, used `mimxrt1050_evk`
- [ ] xmc4xxx
- [x] e1000, used `qemu_x86`
- [x] nxp_s32_netc

### CODE REVIEW

**Comment by jukkar** [comment]
@pdgendt, this should simplify the VLAN support we discussed recently, the code does not yet work fully but I am working on it.

File: subsys/net/l2/ethernet/vlan.c [hunk 1]
```diff
@@ -0,0 +1,697 @@
...
+			if (iface == vlan_ctx[i]->attached_to &&
+			    (vlan_tag == NET_VLAN_TAG_UNSPEC ||
+			     vlan_tag == vlan_ctx[i]->tag)) {
+				return vlan_ctx[i];
+			}
+		} else {
+			if (iface == vlan_ctx[i]->iface &&
+			    (vlan_tag == NET_VLAN_TAG_UNSPEC ||
+			     vlan_tag == vlan_ctx[i]->tag)) {
+				return vlan_ctx[i];
```
Comment (pdgendt) [inline subsys/net/l2/ethernet/vlan.c:157]: `return` without unlock

**Comment by jukkar** [comment]
* Rebased against upstream
* IPv6 over VLAN works
* ARP over VLAN still has some issues
* tests not touched yet
* native_sim driver fixed, all the other drivers pending

**Comment by jukkar** [comment]
* Rebased against upstream
* ARP over VLAN still has some issues. The pending packet that is sent after an ARP reply, is sent without a VLAN tag. This is a bit tricky to fix.
* VLAN test fixed to work with newer VLAN

**Comment by jukkar** [comment]
As the description says, this refactors the VLAN support to use virtual network interfaces we have in zephyr. What this means in practice:
* we use less resources in the system (especially if there are multiple Ethernet network interfaces)
* the VLAN packet handling does not need to be done in the Ethernet driver

### Setup Linux VLAN interfaces for simple testing

If using native_sim board, do this
```
cd tools/net-setup
./net-setup -c zeth-vlan.conf
```
And compile the exe like this
```
west build -p -b native_sim -d ~/zephyrproject/build/echo-server ~/zephyrproject/zephyr/samples/net/sockets/echo_server/ -- \
-DCONFIG_NATIVE_UART_AUTOATTACH_DEFAULT_CMD="\"gnome-terminal -- screen %s\"" \
-DCONFIG_LOG_BUFFER_SIZE=65535 \
-DCONFIG_NET_INTERFACE_NAME_LEN=15 \
-DOVERLAY_CONFIG=overlay-vlan.conf

~/zephyrproject/build/echo-server/zephyr/zephyr.exe -attach_uart
```

Currently supported boards that enable VLAN in driver capabilities:

- nxp_s32_netc
- e1000
- xmc4xxx
- nxp_enet
- nxp_s32_gmac
- mcux
- enc28j60
- sam_gmac
- stm32
- native_sim

This script could be used to compile and flash echo-server to a specific board:
```
#!/bin/bash

if [ "$1" == "" ]; then
    BOARD=native_sim
    OPTS=""
else
    BOARD="$1"
    OPTS=""
fi

west build -p -b $BOARD -d ~/zephyrproject/build/echo-server ~/zephyrproject/zephyr/samples/net/sockets/echo_server/ -- \
-DCONFIG_NATIVE_UART_AUTOATTACH_DEFAULT_CMD="\"gnome-terminal -- screen %s\"" \
-DCONFIG_LOG_BUFFER_SIZE=65535 \
-DCONFIG_NET_INTERFACE_NAME_LEN=15 \
-DOVERLAY_CONFIG=overlay-vlan.conf \
-DCONFIG_NET_DHCPV4=y \
$OPTS

if [ "$BOARD" == native_sim ]; then
    if [ $? -eq 0 ]; then
	~/zephyrproject/build/echo-server/zephyr/zephyr.exe -attach_uart
    fi
else
    west flash -d ~/zephyrproject/build/echo-server
fi
```

Example:
```
./vlan-echo-server nucleo_f767zi
```

The following script can be used in Linux to setup/remove the VLAN interfaces, IP addresses and routes.
```
#!/bin/bash

if [ -z "$1" ]; then
    echo "$0 setup|remove"
    exit 1
fi

ACTION="$1"

if [ -z "$2" ]; then
    IFACE=enp0s31f6
else
    IFACE="$2"
fi

if [ "$ACTION" == setup ]; then
    sudo ip link add link "$IFACE" name vlan.100 type vlan id 100
    sudo ip link set vlan.100 type up
    sudo ip addr add 198.51.100.2/24 dev vlan.100
    sudo ip route add 198.51.100/24 dev vlan.100
    sudo ip -6 addr add 2001:db8:100::2/64 dev vlan.100
    sudo ip -6 route add 2001:db8:100::/64 dev vlan.100
    sudo ifconfig vlan.100 up

    sudo ip link add link "$IFACE" name vlan.200 type vlan id 200
    sudo ip link set vlan.200 type up
    sudo ip addr add 203.0.113.2/24 dev vlan.200
    sudo ip route add 203.0.113/24 dev vlan.200
    sudo ip -6 addr add 2001:db8:200::2/64 dev vlan.200
    sudo ip -6 route add 2001:db8:200::/64 dev vlan.200
    sudo ifconfig vlan.200 up

    exit
fi

if [ "$ACTION" == remove ]; then
    sudo ifconfig vlan.100 down
    sudo ip -6 route del 2001:db8:100::/64 dev vlan.100
    sudo ip -6 addr del 2001:db8:100::2/64 dev vlan.100
    sudo ip route del 198.51.100/24 dev vlan.100
    sudo ip addr del 198.51.100.2/24 dev vlan.100
    sudo ip link set vlan.100 type down
    sudo ip link del link "$IFACE" name vlan.100

    sudo ifconfig vlan.200 down
    sudo ip -6 route del 2001:db8:200::/64 dev vlan.200
    sudo ip -6 addr del 2001:db8:200::2/64 dev vlan.200
    sudo ip route del 203.0.113/24 dev vlan.200
    sudo ip addr del 203.0.113.2/24 dev vlan.200
    sudo ip link set vlan.200 type down
    sudo ip link del link "$IFACE" name vlan.200

    exit
fi
```
Example:
```
./vlan.sh setup eth0
```
will configure host VLANs to use `eth0` as a master interface. The `vlan.100` and `vlan.200` interfaces will use the master interface to send packets.

Then
```
ping -c 1 198.51.100.1
```
will send a ICMP Echo-Request to VLAN-100 interface in zephyr.
And
```
ping -c 1 203.0.113.1
```
will send a ICMP Echo-Request to VLAN-200 interface in zephyr.

If using the `echo-server` sample and the `overlay-vlan.conf` file like in above example, the Zephyr VLAN interfaces are by default down, and you must turn them on before they can be pinged.
```
net iface up 2
net iface up 3
```

The Zephyr interfaces look like this with this example `overlay-vlan.conf`
```
uart:~$ net iface

Interface eth0 (0x20023198) (Ethernet) [1]
===================================
Virtual interfaces attached to this : 2 3 
Link addr : 00:80:E1:EB:47:F7
MTU       : 1500
Flags     : AUTO_START,IPv4,IPv6
Device    : ethernet@40028000 (0x802974c)
Ethernet capabilities supported:
        Virtual LAN
        10 Mbits
        100 Mbits
IPv6 unicast addresses (max 3):
        fe80::280:e1ff:feeb:47f7 autoconf tentative infinite
        2001:db8::1 manual preferred infinite
IPv6 multicast addresses (max 4):
        ff02::1
        ff02::1:ffeb:47f7  <not joined>
        ff02::1:ff00:1
IPv6 prefixes (max 2):
        <none>
IPv6 hop limit           : 64
IPv6 base reachable time : 30000
IPv6 reachable time      : 35562
IPv6 retransmit timer    : 0
IPv4 unicast addresses (max 1):
        192.168.88.14/255.255.255.0 DHCP preferred
IPv4 multicast addresses (max 2):
        224.0.0.1
IPv4 gateway : 192.168.88.1
DHCPv4 lease time : 600
DHCPv4 renew time : 300
DHCPv4 server     : 192.168.88.1
DHCPv4 requested  : 192.168.88.14
DHCPv4 state      : bound
DHCPv4 attempts   : 1

Interface VLAN-100 (0x20023300) (Virtual) [2]
==================================
Virtual name : VLAN to 1
Attached  : 1 (Ethernet / 0x20023198)
Link addr : 00:80:E1:EB:47:F7
MTU       : 1500
Flags     : NO_AUTO_START,IPv4,IPv6
Device    : VLAN_0 (0x8029738)
VLAN tag  : 100 (0x064)
IPv6 unicast addresses (max 3):
        2001:db8:100::1 manual preferred infinite
        fe80::280:e1ff:feeb:47f7 autoconf preferred infinite
IPv6 multicast addresses (max 4):
        ff02::1
        ff02::1:ff00:1
        ff02::1:ffeb:47f7
IPv6 prefixes (max 2):
        <none>
IPv6 hop limit           : 64
IPv6 base reachable time : 30000
IPv6 reachable time      : 22689
IPv6 retransmit timer    : 0
IPv4 unicast addresses (max 1):
        198.51.100.1/255.255.255.0 manual preferred infinite
IPv4 multicast addresses (max 2):
        224.0.0.1
IPv4 gateway : 0.0.0.0
DHCPv4 lease time : 0
DHCPv4 renew time : 0
DHCPv4 server     : 0.0.0.0
DHCPv4 requested  : 0.0.0.0
DHCPv4 state      : disabled
DHCPv4 attempts   : 0

Interface VLAN-200 (0x20023468) (Virtual) [3]
==================================
Virtual name : VLAN to 1
Attached  : 1 (Ethernet / 0x20023198)
Link addr : 00:80:E1:EB:47:F7
MTU       : 1500
Flags     : NO_AUTO_START,IPv4,IPv6
Device    : VLAN_1 (0x8029724)
VLAN tag  : 200 (0x0c8)
IPv6 unicast addresses (max 3):
        2001:db8:200::1 manual preferred infinite
        fe80::280:e1ff:feeb:47f7 autoconf preferred infinite
IPv6 multicast addresses (max 4):
        ff02::1
        ff02::1:ff00:1
        ff02::1:ffeb:47f7
IPv6 prefixes (max 2):
        <none>
IPv6 hop limit           : 64
IPv6 base reachable time : 30000
IPv6 reachable time      : 25054
IPv6 retransmit timer    : 0
IPv4 unicast addresses (max 1):
        203.0.113.1/255.255.255.0 manual preferred infinite
IPv4 multicast addresses (max 2):
        224.0.0.1
IPv4 gateway : 0.0.0.0
DHCPv4 lease time : 0
DHCPv4 renew time : 0
DHCPv4 server     : 0.0.0.0
DHCPv4 requested  : 0.0.0.0
DHCPv4 state      : disabled
DHCPv4 attempts   : 0
```

**Comment by jukkar** [comment]
TODO:
* statistics need to updated for VLAN interfaces
* more testing with real boards (reviewers, please help here :bowing_man:)

**Comment by decsny** [comment]
> TODO:
> 
> * ~statistics need to updated for VLAN interfaces~
> * more testing with real boards (reviewers, please help here 🙇‍♂️)
> * verify that priority of the sent/received packets is set correctly

perhaps create a checklist of drivers that need tested?

File: subsys/net/l2/ethernet/vlan.c [hunk 2]
```diff
@@ -0,0 +1,634 @@
...
+}
+
+static int virt_dev_init(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	init_context_iface();
+
+	return 0;
+}
```
[the remark this answers is not in the document]
Comment (jukkar) [reply subsys/net/l2/ethernet/vlan.c:114]: The `init_context_iface()` will map the VLAN virtual iface to each context when the virtual device is initialized. This avoids figuring out the mapping later when tags are allocated.

**Review by pdgendt — DISMISSED** [review]

**Review by Dat-NguyenDuy — CHANGES_REQUESTED** [review]
samples\net\vlan is working in main but is failing with this PR on s32z270dc2 boards (ethernet netc driver):

image

After tried to increase the `CONFIG_NET_IF_MAX_IPV6_COUNT` and `CONFIG_NET_IF_MAX_IPV4_COUNT` to 3 but it's still not working:

image

 I'll try to debug but in the meantime, please do not merge this PR

**Comment by jukkar** [comment]
> I'll try to debug but in the meantime, please do not merge this PR

@Dat-NguyenDuy 
Can you try with `samples/net/sockets/echo-server` instead? I have instructions how to do that somewhere in this PR.
This way we can rule out the sample application issue?
I can also try with vlan-sample. I have not really used it much because it is not very usable for anything more complex (like TCP / UDP communication).

**Comment by Dat-NguyenDuy** [comment]
> > I'll try to debug but in the meantime, please do not merge this PR
> 
> @Dat-NguyenDuy 
> Can you try with `samples/net/sockets/echo-server` instead? I have instructions how to do that somewhere in this PR.
> This way we can rule out the sample application issue?
> I can also try with vlan-sample. I have not really used it much because it is not very usable for anything more complex (like TCP / UDP communication).

Sure. I will try

**Review by manuargue — DISMISSED** [review]
I've tested `samples/net/sockets/echo_server` on both `nxp_s32_gmac` and `nxp_s32_netc`, all passing:

```
# nxp_s32_gmac
$ sudo ping -c 1000 198.51.100.1 -i 0 | tail -2
1000 packets transmitted, 1000 received, 0% packet loss, time 262ms
rtt min/avg/max/mdev = 0.223/0.235/0.294/0.013 ms, ipg/ewma 0.261/0.230 ms
$ sudo ping -c 1000 203.0.113.1 -i 0 | tail -2
1000 packets transmitted, 1000 received, 0% packet loss, time 264ms
rtt min/avg/max/mdev = 0.227/0.237/1.103/0.029 ms, ipg/ewma 0.264/0.275 ms
```

```
# nxp_s32_netc
$ sudo ping -c 1000 198.51.100.1 -i 0 | tail -2
1000 packets transmitted, 1000 received, 0% packet loss, time 321ms
rtt min/avg/max/mdev = 0.280/0.283/0.597/0.010 ms, ipg/ewma 0.321/0.282 ms
$ sudo ping -c 1000 203.0.113.1 -i 0 | tail -2
1000 packets transmitted, 1000 received, 0% packet loss, time 323ms
rtt min/avg/max/mdev = 0.280/0.285/0.577/0.011 ms, ipg/ewma 0.323/0.282 ms
```

For the `samples/net/sockets/echo_server` on `nxp_s32_netc` I also needed to add `-DCONFIG_NET_L2_ETHERNET=y` because is not automatically selected when `CONFIG_NETWORKING=y`.

**Comment by jukkar** [comment]
> samples\net\vlan is working in main but is failing with this PR on s32z270dc2 boards (ethernet netc driver):

@Dat-NguyenDuy 
There was an error in vlan sample which caused a crash. I will add a commit to this PR that fixes the crash issue.

**Comment by Dat-NguyenDuy** [comment]
> > samples\net\vlan is working in main but is failing with this PR on s32z270dc2 boards (ethernet netc driver):
> 
> @Dat-NguyenDuy There was an error in vlan sample which caused a crash. I will add a commit to this PR that fixes the crash issue.

Thanks for solving the issue quickly

**Comment by jukkar** [comment]
Changes:
* updated to latest main
* went through all the samples that used VLAN and fixed them
* added information to migration guide

@Dat-NguyenDuy the VLAN sample should work now, please take a look.

**Comment by Dat-NguyenDuy** [comment]
> Changes:
> 
> * updated to latest main
> * went through all the samples that used VLAN and fixed them
> * added information to migration guide
> 
> @Dat-NguyenDuy the VLAN sample should work now, please take a look.

Hi, i've tested the sample, but seems still doesn't work:

image

image

image

Do you see the similar issue on any boards?

**Comment by jukkar** [comment]
The interface is down so your ping is dropped.
Try setting `net iface up 2` and then try to ping again.

**Comment by jukkar** [comment]
The error messages can be ignored, they are printed because the interface was down when doing IPv6 DAD. The operations will be re-done after interface is up.

**Comment by Dat-NguyenDuy** [comment]
> The interface is down so your ping is dropped. Try setting `net iface up 2` and then try to ping again.

Thanks for the suggestions. The after bring iface 2 and 3 up, i see only the iface 3 has unicast address then the ping is working with this iface. But not for iface 2.

image

FYI, i need to increase `CONFIG_NET_IF_MAX_IPV4_COUNT` to 3 otherwise there is a warning`You have 2 IPv4 net_if addresses but 3 network interfaces`

File: samples/net/vlan/prj.conf [hunk 3]
```diff
@@ -18,6 +18,9 @@ CONFIG_NET_IF_MCAST_IPV6_ADDR_COUNT=5
 CONFIG_NET_IF_UNICAST_IPV4_ADDR_COUNT=1
 CONFIG_NET_MAX_CONTEXTS=10
 
+CONFIG_NET_IF_MAX_IPV6_COUNT=3
```
Comment (Dat-NguyenDuy) [inline samples/net/vlan/prj.conf:21]: `CONFIG_NET_IF_MAX_IPV4_COUNT=3`

**Comment by Dat-NguyenDuy** [comment]
Perhaps the sample can be updated? I see that after `init_app` function in main finished, the last element in `ipv4_addresses` is not assigned an IP

image

File: samples/net/vlan/prj.conf (hunk 3 above)
Comment (jukkar) [reply samples/net/vlan/prj.conf:21]: Indeed, should have been ipv4 count. Will fix.

**Comment by jukkar** [comment]
> jenkins@CPT00135429C:~$ sudo ping -c 1000 198.51.100.1 -i 0 | tail -2

The VLAN sample is only supporting 203.0.113.1/24 address range set to interface 3. Dunno why I did it like that, I can look into this in a separate PR and configure addresses to interface 2 too. The `CONFIG_NET_IF_MAX_IPV4_COUNT=3` does not change the situation here.

**Comment by jukkar** [comment]
I do not have a boards using `enc28j60` nor `xmc4xxx` drivers so cannot test those. I would assume they work fine like the other drivers have. If not, then we can fix any issues in subsequent PRs.

**Review by Dat-NguyenDuy — APPROVED** [review]

**Review by manuargue — APPROVED** [review]

**Review by pdgendt — APPROVED** [review]

**Review by rlubos — APPROVED** [review]
LGTM

**Comment by go2sh** [comment]
Hey,

this brakes vlan tag striping or at least its not clear to me how this would work. The comment in `ethernet_recv` mentioned, that the VLAN packets should be routed via l2 virtual interface through here to keep packet stats intact. A packet without VLAN header but TCI set through the driver is not processed correctly.

Further the packet priority is not set for VLAN tagged packet.

I have some local fixes for that, which I can submit.
BR,
go2hs

**Comment by fabiobaltieri** [comment]
@go2sh if you have a fix in hand, opening a PR is certainly the quickest way of getting it fixed.

**Comment by go2sh** [comment]
@fabiobaltieri See #73075

**Pull request merged**